iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -1291,329 +1338,409 @@ static bool isSystemProperty(const char* tokenData)
 // such that the application can call IoTHubMessage_GetInputName() to retrieve this.  This is only currently used
 // in IoT Edge module to module message communication, so that this module receiving the message can know which module invoked in.
 //
-static int addInputNamePropertyToMsg(IOTHUB_MESSAGE_HANDLE IoTHubMessage, const char* topic_name)
+// For IoT Edge module to module communication, the incoming topic will be of the form devices/{deviceId}/modules/{moduleId}/inputs/{inputName}.
+// When this function is called, the caller has already skipped past the devices/{deviceId}/modules prefix.  We would start at inputs/{inputName}.
+// On return, we indicate where properties (if specified) start for this message.
+//
+static const char* addInputNamePropertyToMsg(IOTHUB_MESSAGE_HANDLE iotHubMessage, const char* propertiesStart)
 {
-    // Codes_SRS_IOTHUB_TRANSPORT_MQTT_COMMON_31_061: [ If the message is sent to an input queue, `IoTHubTransport_MQTT_Common_DoWork` shall parse out to the input queue name and store it in the message with IoTHubMessage_SetInputName ]
-    // Codes_SRS_IOTHUB_TRANSPORT_MQTT_COMMON_31_062: [ If IoTHubTransport_MQTT_Common_DoWork receives a malformatted inputQueue, it shall fail ]
+    const char* result;
+    const char* inputNameStart;
+    const char* inputNameEnd;
+    char* inputNameCopy = NULL;
 
-    int result = MU_FAILURE;
-    int number_tokens_read = 0;
-
-    STRING_TOKENIZER_HANDLE token_handle = STRING_TOKENIZER_create_from_char(topic_name);
-    if (token_handle == NULL)
+    if (((inputNameStart = strchr(propertiesStart, TOPIC_SLASH)) == NULL) || (*(inputNameStart + 1) == '\0'))
     {
-        LogError("STRING_TOKENIZER_create_from_char failed\n");
-        result = MU_FAILURE;
+        LogError("Cannot find '/' to mark beginning of input name");
+        result = NULL;
     }
     else
     {
-        STRING_HANDLE token_value;
-        if ((token_value = STRING_new()) == NULL)
+        inputNameStart++;
+        if ((inputNameEnd = strchr(inputNameStart, TOPIC_SLASH)) == NULL)
         {
-            LogError("Failed allocating token_value");
+            LogError("Cannot find '/' after input name");
+            result = NULL;
         }
-        else
+        else 
         {
-            while (STRING_TOKENIZER_get_next_token(token_handle, token_value, "/") == 0)
+            size_t inputNameLength = inputNameEnd - inputNameStart;
```

Comment:
Should we care if Edge - in a super rare situation - sent us a topic with empty input name ("//")? In that case `inputNameLength` would be 0 and we would set an empty input name in the message handle. But that would be wrong... But we would also be just conveying what the service (in this case Edge) is giving us.